modules/network/firewall-rules/main.tf

```diff
@@ -14,25 +14,46 @@
  * limitations under the License.
  */
 
+locals {
+  use_subnetwork_data = (var.project_id == null || var.network_name == null) && var.subnetwork_self_link != null
+}
+
 # the google_compute_network data source does not allow identification by
 # self_link, which uniquely identifies subnet, project, and network
 data "google_compute_subnetwork" "subnetwork" {
+  # Only instantiate this data source if needed
+  count     = local.use_subnetwork_data ? 1 : 0
   self_link = var.subnetwork_self_link
 }
 
-# Module-level check for Private Google Access on the subnetwork
-check "private_google_access_enabled_subnetwork" {
-  assert {
-    condition     = data.google_compute_subnetwork.subnetwork.private_ip_google_access
-    error_message = "Private Google Access is disabled for subnetwork '${data.google_compute_subnetwork.subnetwork.name}'. This may cause connectivity issues for instances without external IPs trying to access Google APIs and services."
+locals {
+  # Derived values from data source, null if data source is not used
+  derived_project_id   = local.use_subnetwork_data ? data.google_compute_subnetwork.subnetwork[0].project : null
+  derived_network_name = local.use_subnetwork_data ? data.google_compute_subnetwork.subnetwork[0].network : null
+
+  # Effective values: Use var if provided, otherwise use derived value
+  effective_project_id   = coalesce(var.project_id, local.derived_project_id)
```

Comment:
local.derived_project_id is more assertive than var.project_id as project_id provided could be wrong. I guess we should swap the variable positions